File: app/src/main/java/com/benny/pxerstudio/shape/draw/DrawShape.java
package com.benny.pxerstudio.shape.draw;

import android.graphics.Bitmap;

import androidx.core.graphics.ColorUtils;

import com.benny.pxerstudio.shape.BaseShape;
import com.benny.pxerstudio.widget.PxerView;

import java.util.ArrayList;

public class DrawShape extends BaseShape {

    protected void drawOnLayer(final Bitmap layerToDraw, final PxerView pxerView,
                               final int x, final int y) {
        layerToDraw.setPixel(x, y,
                ColorUtils.compositeColors(
                        pxerView.getSelectedColor(),
                        layerToDraw.getPixel(x, y)));
    }

    protected void addPxerView(final Bitmap layerToDraw, final ArrayList<PxerView.Pxer> previousPxer,
                               final int x, final int y) {
        previousPxer.add(new PxerView.Pxer(x, y, layerToDraw.getPixel(x, y)));
    }

    protected void draw(final Bitmap layerToDraw, final ArrayList<PxerView.Pxer> previousPxer) {
        for (int i = 0; i < previousPxer.size(); i++) {
            final PxerView.Pxer pxer = previousPxer.get(i);
            layerToDraw.setPixel(pxer.getX(), pxer.getY(), pxer.getColor());
        }
        previousPxer.clear();
    }
}

File: app/src/main/java/com/benny/pxerstudio/shape/draw/DrawShape.kt
package com.benny.pxerstudio.shape.draw

import android.graphics.Bitmap
import androidx.core.graphics.ColorUtils
import com.benny.pxerstudio.shape.BaseShape
import com.benny.pxerstudio.widget.PxerView
import java.util.*

open class DrawShape : BaseShape() {
    protected fun drawOnLayer(layerToDraw: Bitmap, pxerView: PxerView, x: Int, y: Int) {
        layerToDraw.setPixel(
            x, y, ColorUtils.compositeColors(
                pxerView.selectedColor,
                layerToDraw.getPixel(x, y)
            )
        )
    }

    protected fun addPxerView(
        layerToDraw: Bitmap,
        previousPxer: ArrayList<PxerView.Pxer>,
        x: Int,
        y: Int
    ) {
        previousPxer.add(PxerView.Pxer(x, y, layerToDraw.getPixel(x, y)))
    }

    protected fun draw(layerToDraw: Bitmap, previousPxer: ArrayList<PxerView.Pxer>) {
        for (i in previousPxer.indices) {
            val (x, y, color) = previousPxer[i]
            layerToDraw.setPixel(x, y, color)
        }
        previousPxer.clear()
    }
}
